### PR TITLE
feat(permissions): Split chat and reaction permissions

### LINKED ISSUE #11329: Split "can post messages and reaction permissions"
### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are interested into the same feature.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

---

**Is your feature request related to a problem? Please describe.**

Im trying to setup a scenario, where a channel is an information/notification channel only (only moderator can post messages), but on top of that I'd like to allow other users to react to those messages.
Currently when below option is disabled, users can't post messages and also can't react to moderator posts.

image

**Describe the solution you'd like**

I'd like to have 2 options instead of current one:
- Can post messages
- Can react to messages

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by luflow**
### ☑️ Resolves

* Fix #11329

## Summary

This PR separates the combined "chat" permission into two distinct permissions:
- **PERMISSIONS_CHAT (128)**: For posting messages
- **PERMISSIONS_REACT (256)**: For adding/removing reactions

This allows admins to create "announcement channels" where only moderators can post messages, but all users can still react to those messages.

### Backend Changes
- Added `PERMISSIONS_REACT = 256` constant to `Attendee` model
- Added `REACT` permission attribute in `RequirePermission`
- Added REACT permission check in `InjectionMiddleware`
- Updated `ReactionController` to use REACT permission instead of CHAT
- Added database migration to grant REACT permission to all users who have CHAT permission
- Fixed hardcoded `255` validation in `RoomService` to use `PERMISSIONS_MAX_CUSTOM`

### Frontend Changes
- Added `REACT: 256` constant to `constants.ts`
- Updated `MAX_DEFAULT` to `510` and `MAX_CUSTOM` to `511`
- Split `PermissionsEditor` checkbox "Can post messages and reactions" into two separate checkboxes:
  - "Can post messages" (CHAT)
  - "Can add reactions" (REACT)
- Updated `MessageItem.vue` to check REACT permission for `canReact`

### Migration Strategy
The migration ensures **backward compatibility**:
- All existing records with `CHAT (128)` permission automatically get `REACT (256)` added
- Old clients continue to work because they ignore unknown permission bits
- No frontend fallbacks needed - the UI will work correctly after data migration

### 🖼️ Screenshots / Screencasts

🏚️ Before | 🏡 After
-- | --
Bildschirmfoto 2026-01-24 um 00 56 34 | Bildschirmfoto 2026-01-24 um 00 53 41

### 🚧 Tasks

- [x] Frontend constants updated
- [x] Frontend UI split into two checkboxes

### 🏁 Checklist

- [ ] 🌏 Tested with different browsers / clients:
  - [x] Chromium (Chrome / Edge / Opera / Brave)
  - [ ] Firefox
  - [ ] Safari
  - [ ] Talk Desktop
  - [ ] Integrations with Files sidebar and other apps
  - [x] Not risky to browser differences / client
- [ ] 🖌️ Design was reviewed, approved or inspired by the design team
- [x] ⛑️ Tests are included or not possible
- [x] 📗 User documentation in https://github.com/nextcloud/documentation/tree/master/user_manual/talk has been updated or is not required

---

### 🚧 Tasks

- [x] Backend permission constant added
- [x] Backend permission check implemented
- [x] New permission constant `PERMISSIONS_REACT = 256` added
- [x] Migration preserves backward compatibility
- [x] Database migration for existing data
- [x] Integration tests updated

### 🏁 Checklist

- [x] ⛑️ Tests (unit and/or integration) are included or not possible
- [x] 📘 API documentation in `docs/` has been updated or is not required
- [x] 🔖 Capability is added or not needed

### CODE REVIEW

File: docs/constants.md [hunk 1]
```diff
@@ -110,7 +110,8 @@
 * `16` Can publish audio stream
 * `32` Can publish video stream
 * `64` Can publish screen sharing stream
-* `128` Can post chat message, share items and do reactions
+* `128` Can post chat message and share items
+* `256` Can add reactions
```
Comment (nickvergessen) [inline docs/constants.md]:
```suggestion
* `256` Can add reactions (New in Nextcloud 34, previously handled by the chat permission)
```

**Comment by nickvergessen** [comment]
We should add an entry to the capabilities, so clients know when to show the new permission and send it to the backend, as well as which permission should be checked when showing the reactions action.

For this we need an edit in:
- lib/Capabilities.php in `public const FEATURES`
- docs/capabilities.md
- And the frontend needs to respect the capability as well, in case you federate with an older Nextcloud or run the desktop client (it reuses the frontend) against an older server

If you need more help with that or we should take over, just let us know

**Comment by nickvergessen** [comment]
Failing sharing tests are unrelated, but `features/callapi/update-call-flags.feature:8` seems to be related:
```diff
-"participantPermissions":254,
+"participantPermissions":510,
```

**Comment by nickvergessen** [comment]
The sharing tests are unrelated and on my todo

**Comment by luflow** [comment]
> We should add an entry to the capabilities, so clients know when to show the new permission and send it to the backend, as well as which permission should be checked when showing the reactions action.

@nickvergessen I was already wondering how we handle this breaking change - capabilities is a good solution for that :) was already tinkering with version numbers… which felt wrong.

what name would you suggest? Something like „react-permission“?

Will create/edit PRs in Android, iOS, desktop clients accordingly.

**Comment by nickvergessen** [comment]
> what name would you suggest? Something like „react-permission“?

Sounds good

> Will create/edit PRs in Android, iOS, desktop clients accordingly.

If you can that's cool, otherwise we raise an issue on the platform.
Desktop is not needed, as it reuses the web frontend from this repo's src/ folder

**Comment by luflow** [comment]
@nickvergessen had some trouble running the tests locally, I hope the test changes are reflecting whats needed now :) Could you rerun the CI?

If anything else is needed, just tell me. As soon as you are happy with the changes I move on to ios&android :)

**Comment by nickvergessen** [comment]
> Could you rerun the CI?

Looks good now (apart from a commit message further up, but we'll ignore that for contributions if rebasing is too painful)

**Comment by luflow** [comment]
> Looks good now (apart from a commit message further up, but we'll ignore that for contributions if rebasing is too painful)

You mean the „apply suggestion“?

**Comment by nickvergessen** [comment]
yes:
grafik

**Comment by nickvergessen** [comment]
- features/federation/permissions.feature:14 seems to have an error when testing against 32
- But we can ignore this for now, I will try to find a way to solve this.

**Comment by luflow** [comment]
> yes: grafik

Fixed!

File: docs/react-permission-integration.md [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+The backend migration automatically grants the REACT permission to all users who previously had the CHAT permission, ensuring backward compatibility.
+
+## UI Changes
+
+The permissions editor should show two separate checkboxes:
+- "Can post messages" (CHAT - 128)
+- "Can add reactions" (REACT - 256)
+
+Instead of the previous combined "Can post messages and reactions" checkbox.
```
Comment (Antreesy) [inline docs/react-permission-integration.md:56]: We should keep the old checkbox for compatibility

File: src/components/MessagesList/MessagesGroup/Message/MessageItem.vue [hunk 3]
```diff
@@ -316,8 +316,13 @@ export default {
 		},
 
 		canReact() {
+			// Fall back to CHAT permission when federating with older servers without react-permission capability
+			const permissionToCheck = hasTalkFeature(this.message.token, 'react-permission')
```
Comment (Antreesy) [inline src/components/MessagesList/MessagesGroup/Message/MessageItem.vue]: I'd feel this name would be more fitting and doesn't require an additional comment
```suggestion
			const hasReactPermissions = hasTalkFeature(this.message.token, 'react-permission')
```

File: docs/react-permission-integration.md (hunk 2 above)
Comment (Antreesy) [reply docs/react-permission-integration.md:56]: See also `src/components/ConversationSettings/ConversationPermissionsSettings.vue` - MAX_DEFAULT and MAX_CUSTOM should respect the capability as well:

```diff
 import IconPencilOutline from 'vue-material-design-icons/PencilOutline.vue'
 import PermissionEditor from '../PermissionsEditor/PermissionsEditor.vue'
 import { PARTICIPANT } from '../../constants.ts'
+import { hasTalkFeature } from '../../services/CapabilitiesManager.ts'
 
 const PERMISSIONS = PARTICIPANT.PERMISSIONS
 
@@ -136,6 +137,30 @@
 			return this.$store.getters.conversation(this.token).defaultPermissions
 		},

+		hasReactPermissions() {
+			return hasTalkFeature(this.token, 'react-permission')
+		}, 

+		/**
+		 * Gives value of the PERMISSIONS.MAX_DEFAULT based on capabilities:
+		 * - legacy: 254
+		 * - with 'react-permissions': 510
+		 */
+		maxDefaultPermission() {
+			if (this.hasReactPermissions) {
+				return PERMISSIONS.MAX_DEFAULT
+			}
+			return 254
+		},
+
+		/**
+		 * Gives value of the PERMISSIONS.MAX_CUSTOM based on capabilities:
+		 * - legacy: 255
+		 * - with 'react-permissions': 511
+		 */
+		maxCustomPermission() {
+			if (this.hasReactPermissions) {
+				return PERMISSIONS.MAX_CUSTOM
+			}
+			return 255
+		},
+
 		/**
 		 * Hides and shows the edit button for advanced permissions.
 		 */
@@ -167,7 +192,7 @@
 			// Compute the permissions value
 			switch (value) {
 				case 'all':
-					permissions = PERMISSIONS.MAX_DEFAULT
+					permissions = this.maxDefaultPermission
 					break
 				case 'restricted':
 					permissions = PERMISSIONS.CALL_JOIN
@@ -207,8 +232,8 @@
 		 */
 		getPermissionRadioValue(value) {
 			switch (value) {
-				case PERMISSIONS.MAX_DEFAULT:
-				case PERMISSIONS.MAX_CUSTOM:
+				case this.maxDefaultPermission:
+				case this.maxCustomPermission:
 					return 'all'
 				case PERMISSIONS.CALL_JOIN:
 				case PERMISSIONS.CALL_JOIN | PERMISSIONS.CUSTOM:
```

I'd prefer to have computed properties written as in example for now, so if we add another permissions, it will be easier to modify

File: src/constants.ts [hunk 4]
```diff
@@ -313,8 +313,9 @@ export const PARTICIPANT = {
 		PUBLISH_VIDEO: 32,
 		PUBLISH_SCREEN: 64,
 		CHAT: 128,
-		MAX_DEFAULT: 254,
-		MAX_CUSTOM: 255,
+		REACT: 256,
+		MAX_DEFAULT: 510,
+		MAX_CUSTOM: 511,
 	},
```
Comment (Antreesy) [inline src/constants.ts:319]: Feel free to also add legacy values to constants, to not leave hardcoded numbers =)
```suggestion
	},
	PERMISSIONS_LEGACY: {
		MAX_DEFAULT_NO_REACT: 254,
		MAX_CUSTOM_NO_REACT: 255,
	},
```

File: src/components/PermissionsEditor/PermissionsEditor.vue [hunk 5]
```diff
@@ -25,9 +25,14 @@
 						{{ t('spreed', 'Skip the lobby') }}
 					</NcCheckboxRadioSwitch>
 					<NcCheckboxRadioSwitch
-						v-model="chatMessagesAndReactions"
+						v-model="chatMessages"
 						class="checkbox">
-						{{ t('spreed', 'Can post messages and reactions') }}
+						{{ t('spreed', 'Can post messages') }}
```
Comment (Antreesy) [inline src/components/PermissionsEditor/PermissionsEditor.vue]: We can keep the new name for the `ref`, just to respect the capability:

```suggestion
						{{ hasReactPermissions ? t('spreed', 'Can post messages') : t('spreed', 'Can post messages and reactions') }}
```

File: src/components/PermissionsEditor/PermissionsEditor.vue [hunk 6]
```diff
@@ -25,9 +25,14 @@
...
+					</NcCheckboxRadioSwitch>
+					<NcCheckboxRadioSwitch
```
Comment (Antreesy) [inline src/components/PermissionsEditor/PermissionsEditor.vue:32]:
```suggestion
					<NcCheckboxRadioSwitch
						v-if="hasReactPermissions"
```

File: src/components/PermissionsEditor/PermissionsEditor.vue
Comment (Antreesy) [inline src/components/PermissionsEditor/PermissionsEditor.vue:1]:
```diff
+import { hasTalkFeature } from '../../services/CapabilitiesManager.ts'
 
+		hasReactPermissions() {
+			// FIXME: token should be passed, but atm we don't have federated moderators,
+			// so can be skipped until refactored
+			return hasTalkFeature('local', 'react-permission')
+		},
+
+		/**
+		 * Gives value of the PERMISSIONS.MAX_DEFAULT based on capabilities:
+		 * - legacy: 254
+		 * - with 'react-permissions': 510
+		 */
+		maxDefaultPermission() {
+			if (this.hasReactPermissions) {
+				return PERMISSIONS.MAX_DEFAULT
+			}
+			return 254
+		},
+
 		permissionsWithDefault() {
 			if (this.permissions !== PERMISSIONS.DEFAULT) {
 				return this.permissions
@@ -178,7 +198,7 @@
 			return loadState(
 				'spreed',
 				'default_permissions',
-				PERMISSIONS.MAX_DEFAULT & ~PERMISSIONS.LOBBY_IGNORE,
+				this.maxDefaultPermission & ~PERMISSIONS.LOBBY_IGNORE,
 			)
 		},
```

File: src/components/PermissionsEditor/PermissionsEditor.vue [hunk 7]
```diff
@@ -184,7 +192,8 @@ export default {
 			return (this.callStart ? PERMISSIONS.CALL_START : 0)
 				| PERMISSIONS.CALL_JOIN // Currently not handled, just adding it, so that manually selecting all checkboxes goes to the "All" permissions state
 				| (this.lobbyIgnore ? PERMISSIONS.LOBBY_IGNORE : 0)
-				| (this.chatMessagesAndReactions ? PERMISSIONS.CHAT : 0)
+				| (this.chatMessages ? PERMISSIONS.CHAT : 0)
+				| (this.chatReactions ? PERMISSIONS.REACT : 0)
```
Comment (Antreesy) [inline src/components/PermissionsEditor/PermissionsEditor.vue]:
```suggestion
				| ((this.hasReactPermissions && this.chatReactions) ? PERMISSIONS.REACT : 0)
```

File: src/components/PermissionsEditor/PermissionsEditor.vue [hunk 8]
```diff
@@ -198,7 +207,8 @@ export default {
 		submitButtonDisabled() {
 			return (!!(this.permissionsWithDefault & PERMISSIONS.CALL_START)) === this.callStart
 				&& !!(this.permissionsWithDefault & PERMISSIONS.LOBBY_IGNORE) === this.lobbyIgnore
-				&& !!(this.permissionsWithDefault & PERMISSIONS.CHAT) === this.chatMessagesAndReactions
+				&& !!(this.permissionsWithDefault & PERMISSIONS.CHAT) === this.chatMessages
+				&& !!(this.permissionsWithDefault & PERMISSIONS.REACT) === this.chatReactions
```
Comment (Antreesy) [inline src/components/PermissionsEditor/PermissionsEditor.vue]:
```suggestion
				&& !!(this.permissionsWithDefault & PERMISSIONS.REACT) === (this.hasReactPermissions && this.chatReactions)
```

File: src/components/PermissionsEditor/PermissionsEditor.vue [hunk 9]
```diff
@@ -220,7 +230,8 @@ export default {
 		writePermissionsToComponent(permissions) {
 			this.callStart = Boolean(permissions & PERMISSIONS.CALL_START)
 			this.lobbyIgnore = Boolean(permissions & PERMISSIONS.LOBBY_IGNORE)
-			this.chatMessagesAndReactions = Boolean(permissions & PERMISSIONS.CHAT)
+			this.chatMessages = Boolean(permissions & PERMISSIONS.CHAT)
+			this.chatReactions = Boolean(permissions & PERMISSIONS.REACT)
```
Comment (Antreesy) [inline src/components/PermissionsEditor/PermissionsEditor.vue]: This would be false with old server, so should be OK. But can also add a `(this.hasReactPermissions && ...)` check here

File: docs/react-permission-integration.md (hunk 2 above)
Comment (nickvergessen) [reply docs/react-permission-integration.md:56]: instead of 254 use `PERMISSIONS.MAX_DEFAULT ~ PERMISSIONS.REACT` or summarize the old permissions, to calculate it? Or is something like that not supported by JS?
PS: I just don't like hardcoded numbers.

Comment (Antreesy) [reply docs/react-permission-integration.md:56]: I didn't think about that, should also work 🙈 
Might be much easier to use `PERMISSIONS.MAX_DEFAULT & ~PERMISSIONS.REACT`, otherwise we can think of possible improvements in follow-up, after this PR is merged

**Comment by Antreesy** [comment]
Forgot to mention during review, but:
- [x] If participants don't have a react permission, they shouldn't be able to click on message reactions (to add a new), and they shouldn't see message actions for it (open Emoji picker from three-dots menu, also quick reactions)
- [x] Opposite, if participants do not have chat permissions, but can react - input field, replies should be locked, but not reactions

**Comment by luflow** [comment]
> Forgot to mention during review, but:
> 
> * [ ]  If participants don't have a react permission, they shouldn't be able to click on message reactions (to add a new), and they shouldn't see message actions for it (open Emoji picker from three-dots menu, also quick reactions)
> * [ ]  Opposite, if participants do not have chat permissions, but can react - input field, replies should be locked, but not reactions

I tested both cases and it worked on my side - where did you see the issue where it did not work?

Or do you mean the addition of the check for „permission exists in this server“ for the desktop app?

**Comment by luflow** [comment]
> Nice work!
> 
> Left some comments regarding frontend compatibility. If you're curious to try it, you can build https://github.com/nextcloud/talk-desktop client locally against your branch and test it with available server.

That means even though it’s bundled with the Nextcloud app the desktop app can diverge? I thought @nickvergessen wrote it’s always rolled out together?

File: src/constants.ts (hunk 4 above)
Comment (luflow) [reply src/constants.ts:319]: Wouldn’t it be better to get the max numbers from the API @nickvergessen @Antreesy 

this looks otherwise like „doomed to stay forever in the code base“

imagine more permissions are added… sounds like hell

**Comment by Antreesy** [comment]
> I tested both cases and it worked on my side - where did you see the issue where it did not work?

My bad for not testing properly then, sorry...

E.g. this button is still clickable and triggers a toast to appear, but there's no server request, so no harm so far. UX can be adjusted in follow-up:
image

> if participants do not have chat permissions, but can react

Weird, but I'm having troubles with sending request to `apps/spreed/api/v4/room/<token>/permissions/default` with anything larger than 255. Will try to fix it locally first

> That means even though it’s bundled with the Nextcloud app the desktop app can diverge? 

Webpage frontend is bundled with backend code. But Desktop app is only using frontend part of it, so I can download the latest app version (with frontend from Talk 22) and connect to old server, that's using Talk 19 - some visual changes may be present, but the core functionality should respect server capabilities - so it should not show a toggle when there's no such permission on connected backend

File: src/components/PermissionsEditor/PermissionsEditor.vue (hunk 8 above)
Comment (luflow) [reply src/components/PermissionsEditor/PermissionsEditor.vue]: fixed

File: docs/react-permission-integration.md (hunk 2 above)
Comment (luflow) [reply docs/react-permission-integration.md:56]: fixed (with api implementation though)

File: src/components/MessagesList/MessagesGroup/Message/MessageItem.vue (hunk 3 above)
Comment (luflow) [reply src/components/MessagesList/MessagesGroup/Message/MessageItem.vue]: fixed

File: src/constants.ts (hunk 4 above)
Comment (luflow) [reply src/constants.ts:319]: fixed (with api implementation though)

File: src/components/PermissionsEditor/PermissionsEditor.vue (hunk 5 above)
Comment (luflow) [reply src/components/PermissionsEditor/PermissionsEditor.vue]: fixed

File: src/components/PermissionsEditor/PermissionsEditor.vue (hunk 7 above)
Comment (luflow) [reply src/components/PermissionsEditor/PermissionsEditor.vue]: fixed

File: src/components/PermissionsEditor/PermissionsEditor.vue (hunk 6 above)
Comment (luflow) [reply src/components/PermissionsEditor/PermissionsEditor.vue:32]: fixed

File: src/components/PermissionsEditor/PermissionsEditor.vue (hunk 9 above)
Comment (luflow) [reply src/components/PermissionsEditor/PermissionsEditor.vue]: fixed

**Comment by luflow** [comment]
@nickvergessen @Antreesy I would say I fixed everything - looking forward to your feedback.

@Antreesy maybe I didn't test it in all edge cases, will do that again. But last time I would say it worked as intended :D I tested it in the web UI on mylocal nextcloud via docker setup

**Comment by luflow** [comment]
@Antreesy found the error - don't know why it worked before in my tests, but the annotations and openai yamls were still having the old limits, thats why the middleware threw a 400 error

Not it works flawlessly. I also fixed the already existing bug that the action menu (send later, etc) and voice button were not disabled when permission "chat" was not given. Now the whole buttons and input row is correctly disabled.

The "react only" also works without issues :)

If there is already a reaction though and you click without permission on the button, the warning "reactions are not permitted" is still shown. I think thats ok? Otherwise we have to make the reactions there also "read only" - I think though its a good ux to have feedback there, when you are normally used to be able to react. Then its actionable for you "aaah, I dont have permission here, good to know"

File: src/components/NewMessage/NewMessageAudioRecorder.vue [hunk 10]
```diff
@@ -311,7 +311,10 @@ export default {
 
 .audio-recorder {
 	display: flex;
-	// Audio record button
+
+	:deep(.button-vue:disabled) {
```
Comment (Antreesy) [inline src/components/NewMessage/NewMessageAudioRecorder.vue]: Let's skip this hotfix, and I'll do it properly upstream instead

File: src/components/PermissionsEditor/PermissionsEditor.vue [hunk 11]
```diff
@@ -220,7 +251,10 @@ export default {
 		writePermissionsToComponent(permissions) {
 			this.callStart = Boolean(permissions & PERMISSIONS.CALL_START)
 			this.lobbyIgnore = Boolean(permissions & PERMISSIONS.LOBBY_IGNORE)
-			this.chatMessagesAndReactions = Boolean(permissions & PERMISSIONS.CHAT)
+			this.chatMessages = Boolean(permissions & PERMISSIONS.CHAT)
+			this.chatReactions = this.hasReactPermissions
+				? Boolean(permissions & PERMISSIONS.REACT)
+				: Boolean(permissions & PERMISSIONS.CHAT)
```
Comment (Antreesy) [inline src/components/PermissionsEditor/PermissionsEditor.vue]: Maybe change it in the if block? Currently if `hasReactPermissions === false`:
- you assign `this.chatReactions = !!PERMISSIONS.CHAT`
- L232 is never truthy
- submit button is never disabled

```suggestion
			if (this.hasReactPermissions) {
				this.chatReactions = Boolean(permissions & PERMISSIONS.REACT)
			}
```

File: src/components/NewMessage/NewMessage.vue [hunk 12]
```diff
@@ -1522,17 +1523,18 @@ export default {
...
 	&:hover,
 	&:focus {
 		background-color: var(--color-background-hover) !important;
 	}
 
-	&:disabled {
+	&:disabled,
+	&[aria-disabled="true"] {
 		opacity: .5 !important;
+		pointer-events: none;
```
Comment (Antreesy) [inline src/components/NewMessage/NewMessage.vue]: Same here for a hotfix

File: tests/integration/features/federation/permissions.feature [hunk 13]
```diff
@@ -40,7 +40,7 @@ Feature: federation/permissions
       | CS          | CS                  |
     Then user "participant3" is participant of room "LOCAL::room" (v4)
       | permissions | attendeePermissions |
-      | SJAVPM      | D                  |
+      | SJAVPMR     | D                  |
```
Comment (nickvergessen) [inline tests/integration/features/federation/permissions.feature:44]: Can you add `@skip33` to the test for now, similarly https://github.com/nextcloud/spreed/blob/57663938d58eb96d048cd5f9975d8948b6ea8fda/tests/integration/features/federation/chat.feature#L560
That should allow CI to pass in the federation case against an older version for now.

**Review by nickvergessen — COMMENTED** [review]
Fine for now, but we need a follow up.

---

Follow up:
- Permission check in middleware does not account for federation
- Your instance is 34 but the host is 33
  - On update of your instance your permissions expend to add reaction permission
  - If the host changes your permission still on 33, it does not know reaction permission and will make you set your permissiions without reaction
  - Options:
    - 🐌 check in middleware if the server knows the permission via capability (very bad performance wise)
    - Do math when getting permissions update and check capabilities there (still not good, but less often)
    - Add the list of known permissions (max_custom number would be enough) to the request, so the instance can judge from there if it contains all numbers or needs to do some healing
- Host updates from 33 to 34 **after** you
  - We should let instances know that the users permissions got modified, but doing that in the migration might not work:
    - Server could be "not reachable" so requests could fail, etc
    - Sending many requests during the update is also not good
    - Next join of the participant should be good enough (we just need to make sure to track and notice the changed permission and store it locally?)
- Host updates from 33 to 34 **before** you
  - Currently the migration would simply change your permissions to now include the reaction permission
  - But what if the moderator took that away from you already?
  - Same as above we can/should heal the permissions when joining the room, that is most likely the most important moment and good enough

File: src/components/NewMessage/NewMessage.vue (hunk 12 above)
Comment (luflow) [reply src/components/NewMessage/NewMessage.vue]: fixed

File: src/components/PermissionsEditor/PermissionsEditor.vue (hunk 11 above)
Comment (luflow) [reply src/components/PermissionsEditor/PermissionsEditor.vue]: fixed

File: src/components/NewMessage/NewMessageAudioRecorder.vue (hunk 10 above)
Comment (luflow) [reply src/components/NewMessage/NewMessageAudioRecorder.vue]: fixed

File: tests/integration/features/federation/permissions.feature (hunk 13 above)
Comment (luflow) [reply tests/integration/features/federation/permissions.feature:44]: done

**Comment by luflow** [comment]
@nickvergessen open-api regenerate did not run correctly locally, now fixed! Also eslint was failing because of this. I hope now the important things are green ^^

**Comment by nickvergessen** [comment]
I added you to our org and would create a guest account for you on our instance to improve the communication options. If you don't mind, send me the email address we should invite to `<my github username>@nextcloud.com`

**Comment by luflow** [comment]
> I added you to our org and would create a guest account for you on our instance to improve the communication options. If you don't mind, send me the email address we should invite to `<my github username>@nextcloud.com`

Sounds good :) Done.

**Comment by nickvergessen** [comment]
I sent an invite, in case you didnt get the email, simply try a password reset on cloud.nextcloud.com with the email you sent me

**Comment by nickvergessen** [comment]
Follow up moved to https://github.com/nextcloud/spreed/issues/16902

**Comment by luflow** [comment]
Work continues in https://github.com/nextcloud/spreed/pull/16920
